smoketest/alertlog_test.go

```diff
@@ -0,0 +1,252 @@
+}
+
+func TestNoImmediateNR(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", false, true, true)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var msg = logs.Alert.RecentEvents.Nodes[0].Message
+	assert.Contains(t, msg, "no immediate rule")
+}
+
+func TestNoOnCallUsers(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", true, true, false)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Equal(t, "No one was on-call", details)
+}
+
+func TestNoEPSteps(t *testing.T) {
+	t.Parallel()
+
+	var sql = makeSQL(false, "SMS", true, false, false)
+	h := harness.NewHarness(t, sql, "add-no-notification-alert-log")
+	defer h.Close()
+
+	// create alert
+	doQL(h, t, makeCreateAlertMut(h), nil)
+	h.Trigger()
+	logs := getLogs(h, t)
+
+	// most recent entry
+	var details = logs.Alert.RecentEvents.Nodes[0].State.Details
+	assert.Equal(t, "No escalation policy steps", details)
+}
+
+func makeSQL(disabledCM bool, cmType string, nr bool, epStep bool, epStepUser bool) string {
+	// start initial sql with user and cm
+	var setupSQL = fmt.Sprintf(`
+		insert into users (id, name, email) 
+		values ({{uuid "user"}}, 'bob', 'joe');
+
+		insert into user_contact_methods (id, user_id, name, type, value, disabled) 
+		values ({{uuid "cm1"}}, {{uuid "user"}}, 'personal', '%s', {{phone "1"}}, %t);
+	`, cmType, disabledCM)
+
+	// add notification rule if specified
+	if nr {
+		setupSQL = setupSQL + `
+			insert into user_notification_rules (user_id, contact_method_id, delay_minutes) 
+			values ({{uuid "user"}}, {{uuid "cm1"}}, 0);
+		`
+	}
+
+	// add ep
+	setupSQL = setupSQL + `
+		insert into escalation_policies (id, name) 
+		values ({{uuid "eid"}}, 'esc policy');
+	`
+
+	// add ep step if specified
+	if epStep {
+		setupSQL = setupSQL + `
+			insert into escalation_policy_steps (id, escalation_policy_id) 
+			values ({{uuid "esid"}}, {{uuid "eid"}});
+		`
+	}
+
+	// add user to ep step if step and stepUser specified
+	if epStep && epStepUser {
+		setupSQL = setupSQL + `
+			insert into escalation_policy_actions (escalation_policy_step_id, user_id) 
+			values ({{uuid "esid"}}, {{uuid "user"}});
+		`
+	}
+
+	// add service
+	setupSQL = setupSQL + `
+		insert into services (id, escalation_policy_id, name) 
+		values ({{uuid "sid"}}, {{uuid "eid"}}, 'service');
+	`
+
+	return setupSQL
+}
+
+type logs struct {
+	Alert struct {
+		RecentEvents struct {
+			Nodes []struct {
+				Message string `json:"message"`
+				State   struct {
+					Details string `json:"details"`
+				} `json:"state"`
+			} `json:"nodes"`
+		} `json:"recentEvents"`
+	} `json:"alert"`
```

Comment:
Tags are redundant and can be removed since we're only decoding -- it's already case-insensitive.